core/src/main/java/com/google/zxing/qrcode/detector/FinderPatternFinder.java

```diff
@@ -604,85 +613,52 @@ private boolean haveMultiplyConfirmedCenters() {
       throw NotFoundException.getNotFoundInstance();
     }
 
-    // Filter outlier possibilities whose module size is too different
-    if (startSize > 3) {
-      // But we can only afford to do so if we have at least 4 possibilities to choose from
-      double totalModuleSize = 0.0;
-      double square = 0.0;
-      for (FinderPattern center : possibleCenters) {
-        float size = center.getEstimatedModuleSize();
-        totalModuleSize += size;
-        square += (double) size * size;
-      }
-      double average = totalModuleSize / startSize;
-      float stdDev = (float) Math.sqrt(square / startSize - average * average);
+    Collections.sort(possibleCenters, new EstimatedModuleComparator());
 
-      Collections.sort(possibleCenters, new FurthestFromAverageComparator((float) average));
+    double distotion = Double.MAX_VALUE;
+    double[] squares = new double[3];
+    FinderPattern[] bestPatterns = new FinderPattern[3];
 
-      float limit = Math.max(0.2f * (float) average, stdDev);
+    for (int i = 0; i < possibleCenters.size() - 2; i++) {
+      for (int j = i + 1; j < possibleCenters.size() - 1; j++) {
+        for (int k = j + 1; k < possibleCenters.size(); k++) {
+          float minModuleSize = possibleCenters.get(i).getEstimatedModuleSize();
+          float maxModuleSize = possibleCenters.get(k).getEstimatedModuleSize();
+          if (maxModuleSize > minModuleSize * 1.4f) {
+            // module size is not similar
+            continue;
+          }
 
-      for (int i = 0; i < possibleCenters.size() && possibleCenters.size() > 3; i++) {
-        FinderPattern pattern = possibleCenters.get(i);
-        if (Math.abs(pattern.getEstimatedModuleSize() - average) > limit) {
-          possibleCenters.remove(i);
-          i--;
+          squares[0] = square(possibleCenters.get(i), possibleCenters.get(j));
+          squares[1] = square(possibleCenters.get(j), possibleCenters.get(k));
+          squares[2] = square(possibleCenters.get(k), possibleCenters.get(i));
+          Arrays.sort(squares);
+
+          double d = Math.abs(squares[2] - 2 * squares[1]) + Math.abs(squares[2] - 2 * squares[0]);
```

Comment:
From the Pythagorean theorem.

Reply:
Yes, I get that, but that's not explaining any of the core of the logic. I understand this is looking for an isosceles triangle with sides a, b, and b, and measuring how much a^2 differs from b^2 + b^2. square[2] is the square of the hypotenuse if anything is here, and you compare to the square of both other sides. But why not just measure `Math.abs(squares[2] - squares[1] - squares[0]`? This is the kind of thing that would be useful to write in a comment.

Reply:
Right.
I will write a comment.

Reply:
I

Reply:
I wrote a simple explanation.
Please let me know if not enough.

Reply:
ah...
Please wait, I will rewrite this comment.

Reply:
I rewrote it.
How is this?

Reply:
I think I still have the same question -- this is almost the same as just comparing c^2 to (a^2 + b^2). Would that be more direct and easier to understand?

Reply:
c^2 - (a^2 + b^2) is satisfied by any right triangle, but we need the _isosceles_ right triangle.
for example, a=1, b=√99, c=10 makes it 0. a=7, b=7, c=10 makes it 2.
The best pattern is clearly the latter, but this formula selects the former.

On the case of |c^2 - 2 * b^2| + |c^2 - 2 * a^2|,
a=1, b=√99, c=10 makes it 100.
a=7, b=7, c=10 makes it 4.

So I wrote:
>          // Since any right triangle satisfies the formula c^2 - b^2 - a^2 = 0,
>          // we need to check both two equal sides separately.

Reply:
Ah got it, that make sense.